microprofile/microprofile-fault-tolerance-2.1/src/main/java/org/wildfly/swarm/ts/microprofile/fault/tolerance/v21/retry/RetryService.java

```diff
@@ -0,0 +1,164 @@
+        int invocationNumber = counter.getAndIncrement();
+        if (invocationNumber % 2 == 1) {
+            throw new Throwable("Simulated Throwable");
+        }
+
+        return CompletableFuture.completedFuture(new MyConnection() {
+            @Override
+            public String getData() {
+                return "Hello from throwable_retryOnMyE_abortOnT, counter: " + invocationNumber;
+            }
+        });
+    }
+
+    // First fail, second pass - there will be 1 retry before success
+    @Retry(retryOn = Throwable.class, abortOn = MyException.class)
+    @Fallback(fallbackMethod = "processFallback")
+    public CompletionStage<MyConnection> throwable_retryOnT_abortOnMyE() throws Throwable {
+        int invocationNumber = counter.getAndIncrement();
+        if (invocationNumber % 2 == 1) {
+            throw new Throwable("Simulated Throwable");
+        }
+
+        return CompletableFuture.completedFuture(new MyConnection() {
+            @Override
+            public String getData() {
+                return "Hello from throwable_retryOnT_abortOnMyE, counter: " + invocationNumber;
+            }
+        });
+    }
+
+    // First call goes through and is recognized by fallback (Throwable is parent of Exception)
+    @Retry(retryOn = MyException.class, abortOn = Throwable.class)
+    @Fallback(fallbackMethod = "processFallback")
+    public CompletionStage<MyConnection> myException_retryOnMyE_abortOnT() throws MyException {
+        int invocationNumber = counter.getAndIncrement();
+        if (invocationNumber % 2 == 1) {
+            throw new MyException("Simulated Exception");
+        }
+
+        return CompletableFuture.completedFuture(new MyConnection() {
+            @Override
+            public String getData() {
+                return "Hello from myException_retryOnMyE_abortOnT, counter: " + invocationNumber;
+            }
+        });
+    }
+
+    // First call goes through and is recognized by fallback(abortOn has higher priority than retryOn)
+    @Retry(retryOn = Throwable.class, abortOn = MyException.class)
+    @Fallback(fallbackMethod = "processFallback")
+    public CompletionStage<MyConnection> myException_retryOnT_abortOnMyE() throws MyException {
+        int invocationNumber = counter.getAndIncrement();
+        if (invocationNumber % 2 == 1) {
+            throw new MyException("Simulated Exception");
+        }
+
+        return CompletableFuture.completedFuture(new MyConnection() {
+            @Override
+            public String getData() {
+                return "Hello from myException_retryOnT_abortOnMyE, counter: " + invocationNumber;
+            }
+        });
+    }
+
+    private CompletionStage<MyConnection> processFallback() {
+        return CompletableFuture.completedFuture(new MyConnection() {
+            @Override
+            public String getData() {
+                return "Fallback Hello";
```

Comment:
One more thing: fallback should also include counter value, and correct value should be checked.